src/directory_listing.rs

```diff
@@ -258,59 +241,43 @@ fn read_dir_entries(
                     tracing::error!(
                         "unable to resolve metadata for `{}` symlink (skipped): {:?}",
                         symlink.display(),
-                        err
+                        err,
                     );
                     continue;
                 }
             };
             if symlink_meta.is_dir() {
-                name_encoded.push('/');
                 dirs_count += 1;
+                (FileType::Directory, None)
             } else {
-                size = Some(meta.len());
                 files_count += 1;
+                (FileType::File, Some(symlink_meta.len()))
             }
         } else {
             continue;
-        }
+        };
+
+        let name_encoded = percent_encode(name.as_encoded_bytes(), PERCENT_ENCODE_SET).to_string();
 
-        let mut uri = None;
         // NOTE: Use relative paths by default independently of
         // the "redirect trailing slash" feature.
         // However, when "redirect trailing slash" is disabled
         // and a request path doesn't contain a trailing slash then
         // entries should contain the "parent/entry-name" as a link format.
         // Otherwise, we just use the "entry-name" as a link (default behavior).
         // Note that in both cases, we add a trailing slash if the entry is a directory.
-        if !base_path.ends_with('/') {
-            let base_path = Path::new(base_path);
-            let parent_dir = base_path.parent().unwrap_or(base_path);
-            let mut base_dir = base_path;
-            if base_path != parent_dir {
-                base_dir = match base_path.strip_prefix(parent_dir) {
-                    Ok(v) => v,
-                    Err(err) => {
-                        tracing::error!(
-                            "unable to strip parent path prefix for `{}` (skipped): {:?}",
-                            base_path.display(),
-                            err
-                        );
-                        continue;
-                    }
-                };
-            }
-
-            let mut base_str = String::new();
-            if !base_dir.starts_with("/") {
-                let base_dir = base_dir.to_str().unwrap_or_default();
-                if !base_dir.is_empty() {
-                    base_str.push_str(base_dir);
-                }
-                base_str.push('/');
-            }
+        let mut uri = if !base_path.ends_with('/') && !base_path.is_empty() {
```

Comment:
Maybe we can express the conditional like this below just for clearness.

```suggestion
        let mut uri = if !base_path.is_empty() && !base_path.ends_with('/') {
```

Reply:
I don’t mind either way. My logic was that checking for the trailing slash was the main check here and `!is_empty()` merely a safeguard that quite frankly should never even happen.

Do you want me to create another PR for this?

Reply:
All fine!